src/index.ts

```diff
@@ -476,138 +661,234 @@ export class Upload extends Pumpify {
+      // We should not assume that the server received all bytes sent in the request.
+      // https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+      const range: string = resp.headers.range;
+      this.offset = Number(range.split('-')[1]) + 1;
+
+      // We should not assume that the server received all bytes sent in the request.
+      // - https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+      const missingBytes = this.numBytesWritten - this.offset;
+      if (missingBytes) {
+        const dataToPrependForResending = this.lastChunkSent.slice(
+          -missingBytes
+        );
+        // As multi-chunk uploads send one chunk per request and pulls one
+        // chunk into the pipeline, prepending the missing bytes back should
+        // be fine for the next request.
+        this.unshiftChunkBuffer(dataToPrependForResending);
+        this.numBytesWritten -= missingBytes;
+        this.lastChunkSent = Buffer.alloc(0);
+      }
 
-    // check if this is the same content uploaded previously. this caches a
-    // slice of the first chunk, then compares it with the first byte of
-    // incoming data
-    if (numBytesWritten === 0) {
-      let cachedFirstChunk = this.get('firstChunk');
-      const firstChunk = chunk.slice(0, 16).valueOf();
+      // continue uploading next chunk
+      this.continueUploading();
+    } else if (!this.isSuccessfulResponse(resp.status)) {
+      const err: ApiError = {
+        code: resp.status,
+        name: 'Upload failed',
+        message: 'Upload failed',
+      };
+      this.destroy(err);
+    } else {
+      // remove the last chunk sent
+      this.lastChunkSent = Buffer.alloc(0);
 
-      if (!cachedFirstChunk) {
-        // This is a new upload. Cache the first chunk.
-        this.set({uri: this.uri, firstChunk});
-      } else {
-        // this continues an upload in progress. check if the bytes are the same
-        cachedFirstChunk = Buffer.from(cachedFirstChunk);
-        const nextChunk = Buffer.from(firstChunk);
-        if (Buffer.compare(cachedFirstChunk, nextChunk) !== 0) {
-          // this data is not the same. start a new upload
-          this.bufferStream!.unshift(chunk);
-          this.bufferStream!.unpipe(this.offsetStream);
-          this.restart();
-          return;
-        }
+      if (resp && resp.data) {
+        resp.data.size = Number(resp.data.size);
       }
+      this.emit('metadata', resp.data);
+      this.deleteConfig();
+
+      // Allow the object (Upload) to continue naturally so the user's
+      // "finish" event fires.
+      this.emit('prepareFinish');
     }
+  }
 
-    let length = chunk.length;
+  /**
+   * Check if this is the same content uploaded previously. This caches a
+   * slice of the first chunk, then compares it with the first byte of
+   * incoming data.
+   *
+   * @returns if the request is ok to continue as-is
+   */
+  private async ensureUploadingSameObject() {
```

Comment:
Is this edge case likely to happen? and what would have happened in the previous implementation of `gcs-resumbable-upload`?

If this is an optimization, it might be nice to do this in follow up work just to simplify this PR.

Reply:
This preserves existing behavior - I don't think this is likely to happen (https://github.com/googleapis/gcs-resumable-upload/blob/main/src/index.ts#L577-L596)